### PR TITLE
Allow getting by ID/URL/Title + CustomURL/Username for channels

### PR DESCRIPTION
**Pull request by Tenpi**
I combined all of the `getXByUrl` methods into just one method. I also made it possible to search for a channel using it's customUrl and username, example:
```ts
//These were invalid before, now they work. 
//You can omit the https and/or www
//Getting a channel by username uses the search api internally
let channel = await youtube.getChannel('Tenpi')
let channel = await youtube.getChannel('youtube.com/c/Tenpi')
let video = await youtube.getVideo('www.youtube.com/watch?v=yexu92rKpjs')

//Getting by title also uses the search api.
let video = await youtube.getVideo('interesting title')
let playlist = await youtube.getPlaylist('interesting title')
//I had to modify the tests, because now these also return a video. 
let video = await youtube.getVideo(' ')
let video = await youtube.getVideo('0')
```
I marked the old `getXByUrl` methods as deprecated since they are no longer needed with these changes.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,3 +8,5 @@ typings/
 .nyc_output/
 *.log
 .coveralls.yml
+.env
+src/temp.ts
```
Comment (jasonhaxstuff) [inline .gitignore]: Remove this

File: package.json [hunk 2]
```diff
@@ -8,7 +8,8 @@
     "test": "mocha --exit --require ts-node/register --require source-map-support/register \"./test/!(oauth).spec.ts\"",
     "coverage": "nyc yarn test && nyc report --reporter=text-lcov | coveralls",
     "docs": "gulp typedoc",
-    "prepublishOnly": "gulp"
+    "prepublishOnly": "gulp",
+    "temp": "tsc && node out/temp.js"
```
Comment (jasonhaxstuff) [inline package.json]: Remove this

File: src/index.ts [hunk 3]
```diff
@@ -82,103 +82,103 @@ export class YouTube {
...
    * @param url The url of the video.
    */
   public getVideoByUrl (url: string) {
     const id = parseUrl(url)
 
-    if (!id.video) {
-      return Promise.reject('Not a valid video url')
-    }
-
+    if (!id.video) return Promise.reject('Not a valid video url')
```
Comment (jasonhaxstuff) [inline src/index.ts]: I don't really like 1-liner if statements, could you convert all of these back?

File: src/index.ts [hunk 4]
```diff
@@ -368,6 +368,27 @@ export class YouTube {
 
     return items
   }
+
+  private getId (input: string, type: string) {
+    if (input.includes('youtube.com') || input.includes('youtu.be')) {
+      const result = parseUrl(input)[`${type}`]
+      if (!result.startsWith('UC') && type === 'channel') {
```
Comment (jasonhaxstuff) [inline src/index.ts]: Add a blank line before this line

File: src/index.ts [hunk 5]
```diff
@@ -368,6 +368,27 @@ export class YouTube {
 
     return items
   }
+
+  private getId (input: string, type: string) {
+    if (input.includes('youtube.com') || input.includes('youtu.be')) {
+      const result = parseUrl(input)[`${type}`]
```
Comment (jasonhaxstuff) [inline src/index.ts]: Just use `type` here, not `${type}`

File: src/index.ts [hunk 6]
```diff
@@ -368,6 +368,27 @@ export class YouTube {
...
+      if (!result.startsWith('UC') && type === 'channel') {
+        return request.api('search', { q: result, type: 'channel', part: 'id' }, this.token, this.tokenType).then((r) => r.items[0].id.channelId)
+      } else {
+        return result
+      }
+    } else {
+      if ((input.length <= 20 && type === 'channel')) {
```
Comment (jasonhaxstuff) [inline src/index.ts]: No need to put double parentheses around this

File: src/index.ts [hunk 7]
```diff
@@ -368,6 +368,27 @@ export class YouTube {
 
     return items
   }
+
+  private getId (input: string, type: string) {
+    if (input.includes('youtube.com') || input.includes('youtu.be')) {
+      const result = parseUrl(input)[`${type}`]
+      if (!result.startsWith('UC') && type === 'channel') {
+        return request.api('search', { q: result, type: 'channel', part: 'id' }, this.token, this.tokenType).then((r) => r.items[0].id.channelId)
```
Comment (jasonhaxstuff) [inline src/index.ts]: `q` could be `input`, not `result`.

Comment (jasonhaxstuff) [reply src/index.ts]: should*

Comment (jasonhaxstuff) [reply src/index.ts]: Also remove the parentheses around `r`, and the same for all other single-parameter lambda functions

Comment (jasonhaxstuff) [reply src/index.ts]: `items[0]` may be undefined, so if it is falsey return undefined. Do this for all of these.

Comment (jasonhaxstuff) [reply src/index.ts]: Actually, I'm wrong. `q` should be `result`.

**Comment by Tenpi** [comment]
Thanks for the review, it's my fault for using a vague name but using `result` was correct in that case. `input` would be the full address like `https://www.youtube.com/username`, and `result` would just be the username from the url. Sorry for confusion, I renamed it to `parsedUrl`

**Comment by jasonhaxstuff** [comment]
@Tenpi - It'll be around 8 hours until I can merge this since I have to wait for my YouTube API quota and I'm about to head to bed. It looks good though, I just need to get the test coverage.

**Pull request merged**